// wc/wc.js
#!/usr/bin/env node
/*jslint nodejs:true */

var
	http = require("http"),
	url = require("url"),
	argv = require("optimist").string("h", "s").argv;

// Input: A text string
// Output: A tuple of line count, word count, character count.
function wc(s) {
	var
		lc = 0,
		wc = 0,
		cc = 0;

	try {
		var lines = s.split("\n");

		if (lines[0] == "") {
			lines = [];
		}

		var words = lines.join(" ").split(" ");

		if (words[0] == "") {
			words = [];
		}

		var
			lc = lines.length,
			wc = words.length,
			cc = s.length;
	}
	catch (e) {}

	return [lc, wc, cc];
}

exports.wc = wc;

function page(req) {
	var
		query = url.parse(req.url, true).query,
		text = query.text || "",
		totals = wc(text);

	return "<!DOCTYPE html>" +
		"<head>" +
		"<title>Word Count</title><meta http-equiv=\"Content-Type\" content=\"text/html;charset=utf-8\" />" +
		"</head>" +
		"<body onload=\"document.getElementById('text').focus();\">" +
		"<style type=\"text/css\" />" +
		"body { text-align: center; }" +
		"a { color: #000; }" +
		"a:visited { color: #000; }" +
		"#logo { font-size: 500%; }" +
		"#wc_form { display: block; margin: 0 auto; width: 400px; }" +
		"label { display: block; float: left; width: 5%; padding-right: 1%; }" +
		"#github { position: fixed; bottom: 1em; right: 1em; }" +
		"</style>" +
		"<h1 id=\"logo\"><a href=\"/\">Word Count</a></h1>" +
		"<div id=\"wc_form\">" +

		"<form action=\"\" method=\"get\">" +

		"<p><textarea id=\"text\" name=\"text\" cols=\"50\" rows=\"15\">" + text + "</textarea></p>" +

		"<p><button type=\"submit\">Count</button></p>" +

		"<p><label>Lines:</label>" + totals[0] + "<br/>" +
		"<label>Words:</label>" + totals[1] + "<br/>" +
		"<label>Characters:</label>" + totals[2] + "</p>" +

		"</div>" +
		"<div id=\"github\"><a href=\"https://github.com/mcandre/node-wc\">GitHub</a></div>" +
		"</body></html";
}

function server() {
	http.createServer(function (req, res) {
		res.writeHead(200, {"Content-Type": "text/html"});
		res.end(page(req));
	}).listen(7125, "localhost");

	console.log("Server running at http://localhost:7125/");
}

exports.server = server;

function commandLine() {
	process.stdin.resume();
	process.stdin.setEncoding("UTF8");

	var text = "";

	process.stdin.on("data", function(chunk) {
		text += chunk;
	});

	process.stdin.on("end", function() {
		var totals = wc(text);
		console.log("Lines: " + totals[0]);
		console.log("Words: " + totals[1]);
		console.log("Characters: " + totals[2]);
	});
}

exports.commandLine = commandLine;

function usage() {
	console.log("wc.js [options]");
	console.log("-s\tServer");
	console.log("-h\tUsage");
}

exports.usage = usage;

function main() {
	if ("h" in argv) {
		usage();
	}
	else if ("s" in argv) {
		server();
	}
	else {
		commandLine();
	}
}

if (!module.parent) { main(); }
